lib/faraday/response/json.rb

```diff
@@ -27,7 +27,22 @@ def process_response(env)
       end
 
       def parse(body)
-        ::JSON.parse(body, @parser_options || {}) unless body.strip.empty?
+        return if body.strip.empty?
+
+        parser_options = @parser_options.dup || {}
+
+        decoder, method_name =
+          if parser_options[:decoder].is_a?(Array) && parser_options[:decoder].size >= 2
+            parser_options[:decoder].slice(0, 2)
+          elsif parser_options[:decoder].respond_to?(:decode)
+            [parser_options[:decoder], :decode]
+          else
+            [::JSON, :parse]
+          end
+
+        parser_options.delete(:decoder)
```

Comment:
I'm worried this might affect performance a bit too much since we're duplicating and mutating hashes (thinking about applications making LOTS of req/sec, this could pile up).

This whole block can be moved into the initialiser (or better, in a method that is called by the initialiser) and cache `decoder` and `method` into instance variables to use here in the parse.